File: src/components/pipeline/S3MountService.ts

import { toast } from "@/hooks/use-toast";
import { S3MountFormValues } from "./types";

export const generateMountCode = (values: S3MountFormValues): string => {
  return `# Mount S3 bucket
access_key = "${values.access_key}"
secret_key = "${values.secret_key}"
encoded_secret_key = secret_key.replace("/", "%2F")
aws_bucket_name = "${values.aws_bucket_name}"
mount_name = "${values.mount_name}"

# Create the mount
dbutils.fs.mount("s3a://%s:%s@%s" % (access_key, encoded_secret_key, aws_bucket_name), "/mnt/%s" % mount_name)

# List files in the mounted bucket
# display(dbutils.fs.ls("/mnt/%s" % mount_name))`;
};

export const deployMountToDatabricks = async (
  values: S3MountFormValues,
  workspaceUrl: string,
  token: string,
): Promise<void> => {
  if (!workspaceUrl || !token) {
    toast({
      title: "Deployment failed",
      description: "Missing Databricks workspace URL or access token",
      variant: "destructive",
    });
    return;
  }

  try {
    // Generate notebook content for S3 mounting
    const notebookContent = generateMountCode(values);
    
    // Create a notebook name for the S3 mount
    const notebookName = `S3_Mount_${values.aws_bucket_name}_${Date.now()}`;
    
    // Create the notebook in Databricks workspace
    const response = await fetch(`${workspaceUrl}/api/2.0/workspace/import`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        path: `/Users/${notebookName}`,
        format: 'SOURCE',
        language: 'PYTHON',
        content: Buffer.from(notebookContent).toString('base64'),
        overwrite: true
      }),
    });

    if (response.ok) {
      toast({
        title: "S3 Mount notebook deployed",
        description: `Created notebook: ${notebookName}`,
      });
    } else {
      const errorData = await response.json();
      throw new Error(errorData.message || 'Failed to deploy S3 mount notebook');
    }
  } catch (error) {
    console.error('Error deploying S3 mount notebook:', error);
    toast({
      title: "Deployment failed",
      description: error instanceof Error ? error.message : 'Unknown error',
      variant: "destructive",
    });
  }
};

File: python/services/s3_mount_service.py
import base64
import requests
import time

def generate_mount_code(values):
    """Generate code for mounting S3 bucket in Databricks"""
    # Use bucket name as mount name if mount_name is empty
    effective_mount_name = values.get('mount_name') or values.get('aws_bucket_name')
    
    code = f"""access_key = "{values.get('access_key')}"
secret_key = "{values.get('secret_key')}"
encoded_secret_key = secret_key.replace("/", "%2F")
aws_bucket_name = "{values.get('aws_bucket_name')}"
mount_name = "{effective_mount_name}"

dbutils.fs.mount("s3a://%s:%s@%s" % (access_key, encoded_secret_key, aws_bucket_name), "/mnt/%s" % mount_name)"""

    # Add display command based on user preference
    if values.get('show_display_command'):
        code += "\n\n# Display mounted files\ndisplay(dbutils.fs.ls('/mnt/' + mount_name))"
        
    return code

def deploy_mount_notebook(values, workspace_url, token):
    """Deploy an S3 mount notebook to Databricks"""
    try:
        # Generate notebook content for S3 mounting
        notebook_content = generate_mount_code(values)
        
        # Create a notebook name for the S3 mount
        notebook_name = f"S3_Mount_{values.get('aws_bucket_name')}_{int(time.time())}"
        
        # Create the notebook in Databricks workspace
        response = requests.post(
            f"{workspace_url}/api/2.0/workspace/import",
            headers={
                'Authorization': f'Bearer {token}',
                'Content-Type': 'application/json',
            },
            json={
                'path': f'/Users/{notebook_name}',
                'format': 'SOURCE',
                'language': 'PYTHON',
                'content': base64.b64encode(notebook_content.encode()).decode(),
                'overwrite': True
            }
        )
        
        if response.status_code == 200:
            return {"success": True, "message": f"Mount notebook {notebook_name} created successfully"}
        else:
            return {"success": False, "message": f"Failed to create mount notebook: {response.text}"}
            
    except Exception as e:
        return {"success": False, "message": f"Error: {str(e)}"}